File: common/src/main/java/com/ultreon/mods/advanceddebug/client/menu/pages/EntityPage.java
package com.ultreon.mods.advanceddebug.client.menu.pages;

import com.ultreon.mods.advanceddebug.api.client.menu.DebugPage;
import com.ultreon.mods.advanceddebug.api.client.menu.IDebugRenderContext;
import com.ultreon.mods.advanceddebug.util.TargetUtils;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.GuiGraphics;
import net.minecraft.network.chat.Component;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.EntityType;
import net.minecraft.world.entity.LivingEntity;
import net.minecraft.world.entity.item.ItemEntity;
import net.minecraft.world.phys.EntityHitResult;
import org.jetbrains.annotations.NotNull;

import static net.minecraft.ChatFormatting.GRAY;
import static net.minecraft.ChatFormatting.RED;

public class EntityPage extends DebugPage {
    private final Minecraft mc = Minecraft.getInstance();

    public EntityPage() {
    }

    @Override
    public void render(@NotNull GuiGraphics gfx, IDebugRenderContext ctx) {
        EntityHitResult entityHit = TargetUtils.entity();
        if (entityHit != null) {
            Entity entity = entityHit.getEntity();
            EntityType<? extends Entity> type = entity.getType();
            Component customName = entity.getCustomName();

            ctx.left(GRAY + "Entity Type");
            ctx.left("Height", type.getHeight());
            ctx.left("Name", type.getDescription().getString());
            ctx.left("Size", getSize(type.getDimensions().width, type.getDimensions().height));
            ctx.left("Internal Name", type.arch$registryName());
            ctx.left();

            ctx.left(GRAY + "Entity");
            ctx.left("Air", entity.getAirSupply());
            ctx.left("Max Air", entity.getMaxAirSupply());
            ctx.left("Eye Height", entity.getEyeHeight());
            ctx.left("Look Angle", entity.getLookAngle());
            ctx.left("Riding Entity", entity.getVehicle());
            ctx.left("Riding Offset", entity.getMyRidingOffset());
            ctx.left("Entity UUID", entity.getUUID());
            ctx.left("Entity ID", entity.getId());
            ctx.left("Entity Name", entity.getName().getString());
            ctx.left("Custom Name", customName == null ? null : customName.getString());
            ctx.left("Nametag Visible", entity.isCustomNameVisible());
            ctx.left("Silent", entity.isSilent());
            ctx.left("Attackable", entity.isAttackable());
            ctx.left("Invulnerable", entity.isInvulnerable());
            ctx.left("Invisible", entity.isInvisible());
            ctx.left("Sneaking", entity.isCrouching());
            ctx.left("Sprinting", entity.isSprinting());
            ctx.left("Swimming", entity.isSwimming());
            ctx.left("Pushable", entity.isPushable());
            ctx.left("Pushed By Fluid", entity.isPushedByFluid());

            if (entity instanceof LivingEntity living) {
                ctx.right(GRAY + "Living Entity");
                ctx.right("Health", living.getHealth());
                ctx.right("Max. Health", living.getMaxHealth());
                ctx.right("Main Hand", living.getItemInHand(InteractionHand.MAIN_HAND));
                ctx.right("Off Hand", living.getItemInHand(InteractionHand.OFF_HAND));
                ctx.right("Arrow Count", living.getArrowCount());
                ctx.right("Absorption Amount", living.getAbsorptionAmount());
                ctx.right("Speed", living.getSpeed());
                ctx.right("Last Used Hand", living.getUsedItemHand());
                ctx.right("Attacking Entity", living.getKillCredit());
                ctx.right("Item Usage Ticks Remaining", living.getUseItemRemainingTicks());
                ctx.right("Last Attacked Mob", living.getLastHurtMob());
                ctx.right("Rope Position", living.getRopeHoldPosition(mc.getFrameTime()));
                ctx.right("Pose", living.getPose());
                ctx.right("Scale", living.getScale());
                ctx.right("Armor Value", living.getArmorValue());
            } else if (entity instanceof ItemEntity itemEntity) {
                ctx.right(GRAY + "Item Entity");
                ctx.right("Owner", itemEntity.getOwner());
                ctx.right("Item", itemEntity.getItem());
                ctx.right("Spin", itemEntity.getSpin(minecraft.getFrameTime()));
                ctx.right("Age", itemEntity.getAge());
            }
        } else {
            // not looking at a block, or too far away from one to tell
            ctx.top(RED + "<No Entity Was Found>");
        }
    }
}

File: common/src/main/kotlin/com/ultreon/mods/advanceddebug/client/menu/pages/EntityPage.kt
package com.ultreon.mods.advanceddebug.client.menu.pages

import com.mojang.blaze3d.vertex.PoseStack
import com.ultreon.mods.advanceddebug.api.client.menu.DebugPage
import com.ultreon.mods.advanceddebug.api.client.menu.IDebugRenderContext
import com.ultreon.mods.advanceddebug.api.common.size
import com.ultreon.mods.advanceddebug.util.TargetUtils
import com.ultreon.mods.advanceddebug.util.registryName
import net.minecraft.ChatFormatting
import net.minecraft.client.Minecraft
import net.minecraft.world.InteractionHand
import net.minecraft.world.entity.LivingEntity
import net.minecraft.world.entity.item.ItemEntity

open class EntityPage(modId: String, name: String) : DebugPage(modId, name) {
    private val mc = Minecraft.getInstance()
    override fun render(poseStack: PoseStack, ctx: IDebugRenderContext) {
        val entityHit = TargetUtils.entity()
        if (entityHit != null) {
            val entity = entityHit.entity
            val type = entity.type
            val customName = entity.customName
            ctx.left(ChatFormatting.GRAY.toString() + "Entity Type")
            ctx.left("Height", type.height)
            ctx.left("Name", type.description.string)
            ctx.left("Size", size(type.dimensions.width, type.dimensions.height))
            ctx.left("Internal Name", type.registryName)
            ctx.left()
            ctx.left(ChatFormatting.GRAY.toString() + "Entity")
            ctx.left("Air", entity.airSupply)
            ctx.left("Max Air", entity.maxAirSupply)
            ctx.left("Eye Height", entity.eyeHeight)
            ctx.left("Look Angle", entity.lookAngle)
            ctx.left("Riding Entity", entity.vehicle)
            ctx.left("Riding Offset", entity.myRidingOffset)
            ctx.left("Entity UUID", entity.uuid)
            ctx.left("Entity ID", entity.id)
            ctx.left("Entity Name", entity.name.string)
            ctx.left("Custom Name", customName?.string)
            ctx.left("Nametag Visible", entity.isCustomNameVisible)
            ctx.left("Silent", entity.isSilent)
            ctx.left("Attackable", entity.isAttackable)
            ctx.left("Invulnerable", entity.isInvulnerable)
            ctx.left("Invisible", entity.isInvisible)
            ctx.left("Sneaking", entity.isCrouching)
            ctx.left("Sprinting", entity.isSprinting)
            ctx.left("Swimming", entity.isSwimming)
            ctx.left("Pushable", entity.isPushable)
            ctx.left("Pushed By Fluid", entity.isPushedByFluid)
            if (entity is LivingEntity) {
                ctx.right(ChatFormatting.GRAY.toString() + "Living Entity")
                ctx.right("Health", entity.health)
                ctx.right("Max. Health", entity.maxHealth)
                ctx.right("Main Hand", entity.getItemInHand(InteractionHand.MAIN_HAND))
                ctx.right("Off Hand", entity.getItemInHand(InteractionHand.OFF_HAND))
                ctx.right("Arrow Count", entity.arrowCount)
                ctx.right("Absorption Amount", entity.absorptionAmount)
                ctx.right("Speed", entity.speed)
                ctx.right("Last Used Hand", entity.usedItemHand)
                ctx.right("Attacking Entity", entity.killCredit)
                ctx.right("Item Usage Ticks Remaining", entity.useItemRemainingTicks)
                ctx.right("Last Attacked Mob", entity.lastHurtMob)
                ctx.right("Rope Position", entity.getRopeHoldPosition(minecraft.frameTime))
                ctx.right("Pose", entity.getPose())
                ctx.right("Scale", entity.scale)
                ctx.right("Armor Value", entity.armorValue)
            } else if (entity is ItemEntity) {
                ctx.right(ChatFormatting.GRAY.toString() + "Item Entity")
                ctx.right("Age", entity.age)
                ctx.right("Item", entity.item)
                ctx.right("Owner", entity.owner)
                ctx.right("Thrower", entity.thrower)
                ctx.right("Pose", entity.getPose())
            }
        } else {
            // not looking at a block, or too far away from one to tell
            ctx.top(ChatFormatting.RED.toString() + "<No Entity Was Found>")
        }
    }
}